drivers/spi/spi_silabs_eusart.c

```diff
@@ -252,28 +658,61 @@ static int spi_silabs_eusart_transceive(const struct device *dev, const struct s
 	spi_context_lock(&data->ctx, false, NULL, NULL, config);
 
 	ret = spi_silabs_eusart_configure(dev, config, &control);
-	if (ret < 0) {
+	if (ret) {
 		spi_context_release(&data->ctx, ret);
 		return ret;
 	}
 
 	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
-	spi_silabs_eusart_xfer(dev, config);
+	ret = spi_silabs_eusart_xfer(dev, config);
+
+	if (ret) {
+		spi_context_release(&data->ctx, ret);
+		return ret;
+	}
+#if CONFIG_SPI_ASYNC
```

Comment:
I don't fully understand the purpose of this guard. `spi_context_wait_for_completion` will do the right thing and either wait using a semaphore (which gets posted by `spi_context_complete`) or immediately return if `ctx->asynchronous` is true. And that variable gets set by the argument to `spi_context_lock`.

There seems to in general be unneeded complexity between the implementations of sync and async transfers here -- most other implementations simply implement a single transceive function that takes a `bool async`.